libprobe/src/probe_libc.c

```diff
@@ -2,30 +2,532 @@
+        int_fast16_t diff = ((int_fast16_t)c1[i]) - ((int_fast16_t)c2[i]);
+        if (diff) {
+            return diff;
+        }
+    }
+
+    return 0;
+}
+
+void* probe_libc_memcpy(void* dest, const void* src, size_t n) {
+    if (dest == NULL || n == 0) {
+        return dest;
+    }
+
+    // compiler optimization hint that dest and src may not alias
+    // (as specified by man memcpy)
+    if (dest == src) {
+        __builtin_unreachable();
+    }
+
+    size_t i = 0;
+    for (; (i + sizeof(size_t)) < n; i += sizeof(size_t)) {
+        *((size_t*)(((uintptr_t)dest) + i)) = *((size_t*)(((uintptr_t)src) + i));
+    }
+    for (; i < n; ++i) {
+        ((unsigned char*)dest)[i] = ((unsigned char*)src)[i];
+    }
+
+    return dest;
+}
+
+void* probe_libc_memset(void* s, int c, size_t n) {
+    // OPTIMIZE: vectorize this by generating a size_t of repeated bytes and
+    // copying that
+    for (size_t i = 0; i < n; ++i) {
+        ((unsigned char*)s)[i] = (unsigned char)c;
+    }
+
+    return s;
+}
+
+result_str probe_libc_getcwd(char* buf, size_t size) {
+    int retval = probe_syscall2(SYS_getcwd, (uintptr_t)buf, size);
+    if (retval >= 0) {
+
+        // linux may return the string "(unreachable)" under weird
+        // circumstances (see getcwd(2) -> VERSIONS -> C library/kernel
+        // differences), which without a check might be interpreted as a
+        // relative path, breaking the API contract of getcwd; why does it not
+        // just return an error code like all the other edgecases?
+        // it would seem only god and linus knows
+        if (buf[0] != '/') {
+            return (result_str)ERR(ENOENT);
+        }
+
+        return (result_str)OK(buf);
+    }
+    return (result_str)ERR(-retval);
+}
+
+pid_t probe_libc_getpid(void) { return probe_syscall0(SYS_getpid); }
+
+pid_t probe_libc_getppid(void) { return probe_syscall0(SYS_getppid); }
+
+pid_t probe_libc_gettid(void) { return probe_syscall0(SYS_gettid); }
+
+result_int probe_libc_dup(int oldfd) {
+    int retval = probe_syscall1(SYS_dup, oldfd);
+    if (retval >= 0) {
+        return (result_int)OK(retval);
+    }
+    return (result_int)ERR(-retval);
+}
+
+result_int probe_libc_open(const char* path, int flags, mode_t mode) {
+    ssize_t retval = probe_syscall3(SYS_open, (uintptr_t)path, flags, mode);
+    if (retval >= 0) {
+        return (result_int)OK(retval);
+    }
+    return (result_int)ERR(-retval);
+}
+
+result_int probe_libc_openat(int dirfd, const char* path, int flags, mode_t mode) {
+    ssize_t retval = probe_syscall4(SYS_openat, dirfd, (uintptr_t)path, flags, mode);
+    if (retval >= 0) {
+        return (result_int)OK(retval);
+    }
+    return (result_int)ERR(-retval);
+}
+
+result probe_libc_close(int fd) {
+    ssize_t retval = probe_syscall1(SYS_close, fd);
+    if (retval < 0) {
+        return -retval;
+    }
+    return 0;
+}
+
+result_ssize_t probe_libc_read(int fd, void* buf, size_t count) {
```

Comment:
It's fine to have `probe_libc_read` and `probe_libc_write` should actually not be exposed; we should expose/use higher-level functions like `probe_libc_read_all`.

`util.c`'s `write_bytes` should probably become a `probe_libc` function.